src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarReader.cs

```diff
@@ -57,11 +57,16 @@ public void Dispose()
             {
                 _isDisposed = true;
 
-                if (!_leaveOpen && _dataStreamsToDispose?.Count > 0)
+                if (!_leaveOpen)
                 {
-                    foreach (Stream s in _dataStreamsToDispose)
+                    _archiveStream.Dispose();
```

Comment:
What's the relationship between the streams in _dataStreamsToDispose and _archiveStream?  Do the former streams wrap the latter?  I'm wondering if the archive stream should actually be disposed of after rather than before the _dataStreamsToDispose.

Reply:
The code change is correct: The archive stream represents the whole tar archive, while the data streams represent sub-sections of the archive, each wrapping only the data section of an individual entry.

This means that all the individual data streams need to get disposed first, and the archive stream needs to be disposed at the end.

Reply:
> This means that all the individual data streams need to get disposed first, and the archive stream needs to be disposed at the end.

This is the opposite of what I did here. @carlossanlop shall we fix it?

Reply:
Yes.